lib/src/signed_video_h26x_auth.c

```diff
@@ -128,9 +160,30 @@ verify_gop_hash(signed_video_t *self)
   return (memcmp(gop_info->computed_gop_hash, self->received_gop_hash, hash_size) == 0);
 }
 
-/**
- * Adds the hashes of NALUs used in computing the GOP hash and marks these NALUs as part of the GOP
- * hash.
+/*
+ * Iterates through the NALU list to find the first NALU used in the GOP hash.
+ * If the linked hash has not yet been updated with this NALU's hash, it updates the
+ * linked hash with the first NALU hash and marks it as used.
+ */
+void
+update_link_hash_for_auth(signed_video_t *self, h26x_nalu_list_t *nalu_list)
```

Comment:
You use this function in 3 places; `prepare_for_link_and_gop_hash_verification(...)`, `verify_hashes_with_hash_list(...)` and `verify_hashes_without_sei(...)`.

The function `prepare_for_link_and_gop_hash_verification(...)` is called in `prepare_for_validation(...)`, but only if there is a SEI. That will not happen when the stream is for example not signed. Therefore, you had to add a call in `verify_hashes_without_sei(...)` to cover cases where the SEI is delayed.
Sometimes, `verify_hashes_without_sei(...)` is also called when for example the signature could not be verified. Then, link hash is updated twice, which then made it necessary to add a flag, so the same hash is not added again.
I might be very wrong when I interpret the code, but that is how it looks to me. Wouldn't it be more clean to make sure there is only one update made. For example, you can make `prepare_for_link_and_gop_hash_verification(...)` work also if there is no SEI, or you can break out the link update part from `prepare_for_link_and_gop_hash_verification(...)` and put it directly in `prepare_for_validation(...)`.

Why it is needed to call this update in `verify_hashes_with_hash_list(...)` I haven't figured out yet.

Reply:
Now its only used in maybe_validate_gop